pyfifotax/utils.py

```diff
@@ -0,0 +1,667 @@
+            "quantity": np.float64,
+            "sell_price": np.float64,
+            "fees": np.float64,
+            "currency": str,
+        }
+        df_sell_orders = pd.read_excel(
+            xls, sheet_name="sales", parse_dates=["date"], dtype=dtypes
+        )
+        df_sell_orders["comment"] = ""
+
+        dtypes = {
+            "date": None,
+            "net_amount": np.float64,
+            "fees": np.float64,
+            "currency": str,
+        }
+        df_currency_conversions = pd.read_excel(
+            xls, sheet_name=forex_sheet, parse_dates=["date"], dtype=dtypes
+        )
+        df_currency_conversions["target_currency"] = "EUR"
+        df_currency_conversions["foreign_amount"] = (
+            df_currency_conversions.net_amount + df_currency_conversions.fees
+        )
+        df_currency_conversions.rename(
+            mapper={"fees": "source_fees", "currency": "source_currency"},
+            axis="columns",
+            inplace=True,
+        )
+        df_currency_conversions.drop(
+            labels=["net_amount"], axis="columns", inplace=True
+        )
+        df_currency_conversions["comment"] = ""
+
+        df_stock_splits = None
+        df_espp = None
+        df_currency_movements = None
+
+    return RawData(
+        df_deposits,
+        df_espp,
+        df_dividends,
+        df_buy_orders,
+        df_sell_orders,
+        df_currency_conversions,
+        df_currency_movements,
+        df_stock_splits,
+    )
+
+
+def read_data(sub_dir, file_name):
+    # read list of deposits, sales, dividend payments and currency conversions
+    # sort them to ensure that they are in chronological order
+    file_path = os.path.join(sub_dir, file_name)
+    with pd.ExcelFile(file_path) as xls:
+        dtypes = RSURow.type_dict()
+        df_rsu = pd.read_excel(
+            xls, sheet_name="rsu", parse_dates=["date"], dtype=dtypes
+        )
+
+        dtypes = DividendRow.type_dict()
+        dtypes["date"] = None
+        df_dividends = pd.read_excel(
+            xls, sheet_name="dividends", parse_dates=["date"], dtype=dtypes
+        )
+
+        dtypes = BuyOrderRow.type_dict()
+        dtypes["date"] = None
+        df_buy_orders = pd.read_excel(
+            xls, sheet_name="buy_orders", parse_dates=["date"], dtype=dtypes
+        )
+
+        dtypes = SellOrderRow.type_dict()
+        dtypes["date"] = None
+        df_sell_orders = pd.read_excel(
+            xls, sheet_name="sell_orders", parse_dates=["date"], dtype=dtypes
+        )
+
+        dtypes = CurrencyConversionRow.type_dict()
+        dtypes["date"] = None
+        df_currency_conversions = pd.read_excel(
+            xls, sheet_name="currency_conversions", parse_dates=["date"], dtype=dtypes
+        )
+        df_stock_splits = None  # set later
+
+        dtypes = ESPPRow.type_dict()
+        dtypes["date"] = None
+        df_espp = pd.read_excel(
+            xls, sheet_name="espp", parse_dates=["date"], dtype=dtypes
```

Comment:
Exception with:
```
"dtype '<class 'NoneType'>' not understood (sheet: espp)"
```

In both cases the sheets are empty.

Reply:
fixed, `ESPPRow` had an invalid `default_dict` method which had `None` for all keys throwing this exception